File: Design-Patterns/Bridge/Cpp/character.cpp
#include <iostream>
#include "character.h"

character::character(int id, std::string name, int health, int mana)
{
	id_ = id;
	name_ = name;
	health_ = health;
	mana_ = mana;
	spell_ = std::make_shared<default_spell>(10, 10, 10);
}

character::~character() = default;

void character::set_spell(std::shared_ptr<spell> spell)
{
	if (spell == nullptr)
	{
		throw std::exception("Spell cannot be set to null");
	}

	spell_ = spell;
}

void character::cast_spell()
{
	if (spell_ == nullptr) return;
	if (mana_ < spell_->cost) return;

	std::cout << name_ + " casts a spell." << std::endl;
	spell_->cast();
	mana_ -= spell_->cost;
	if (mana_ < 0) mana_ = 0;
}

void character::take_damage(int value)
{
	health_ -= value;
	if (health_ < 0) health_ = 0;
	std::cout << name_ + " have " + std::to_string(health_) + " health points after taking " + std::to_string(value) + " damage points" << std::endl;
}

File: Design-Patterns/Bridge/CSharp/Bridge/Characters/Character.cs
using System;
using Bridge.Spells;

namespace Bridge.Characters
{
    public abstract class Character
    {
        public int Id { get; protected set; }
        public string Name { get; protected set; }
        public int Health { get; protected set; }
        public int Mana { get; protected set; }
        public Spell Spell { get; protected set; }

        protected Character(int id, string name, int health, int mana)
        {
            Id = id;
            Name = name;
            Health = health;
            Mana = mana;
            Spell = new DefaultSpell(10, 10, 10);
        }

        public virtual void SetSpell(Spell spell)
        {
            if (spell == null) throw new ArgumentNullException("Spell cannot be set to null");

            Spell = spell;
        }

        public virtual void CastSpell()
        {
            if (Spell == null) return;
            if (Mana < Spell.Cost) return;

            Console.WriteLine($"{Name} casts a spell.");
            Spell.Cast();
            Mana -= Spell.Cost;
            if (Mana < 0) Mana = 0;            
        }

        public virtual void TakeDamage(int value)
        {
            Health -= value;
            if (Health < 0) Health = 0;
            Console.WriteLine($"{Name} have {Health} health points after taking {value} damage points");
        }
    }
}
